test/python/transpiler/aqc/test_gradient.py

```diff
@@ -0,0 +1,129 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Tests analytical gradient vs the one computed via finite differences.
+"""
+
+import unittest
+
+import numpy as np
+from scipy.stats import unitary_group
+
+from qiskit.test import QiskitTestCase
+from qiskit.transpiler.synthesis.aqc.cnot_structures import make_cnot_network
+from qiskit.transpiler.synthesis.aqc.cnot_unit_objective import DefaultCNOTUnitObjective
+
+
+class TestGradientAgainstFiniteDiff(QiskitTestCase):
+    """
+    Compares analytical gradient vs the one computed via finite difference
+    approximation. Also, the test demonstrates that the difference between
+    analytical and numerical gradients is up to quadratic term in Taylor
+    expansion for small deltas.
+    """
+
+    def test_gradient(self):
+        """
+        Gradient test for specified number of qubits and circuit depth.
+        """
+        num_qubits = 3
+        num_cnots = 14
+
+        cnots = make_cnot_network(
+            num_qubits=num_qubits, network_layout="spin", connectivity_type="full", depth=num_cnots
+        )
+
+        # Generate random target matrix and random starting point. Repeat until
+        # sufficiently large gradient has been encountered.
+        # This operation is pretty fast.
+        while True:
+            target_matrix = self._random_special_unitary(num_qubits=num_qubits)
```

Comment:
would it be possible to use the existing `qiskit.quantum_info.operators.random.random_unitary` instead of `_random_special_unitary`?

Reply:
A direct replacement is not applicable here, `random_unitary` must be normalized. and there's an additional overhead of the `Operator` class. Anyway, this code is deleted now, since the matrix in the test is fixed now.